File: src/main/java/fr/uiytt/ageofempire/listeners/GameListener.java
package fr.uiytt.ageofempire.listeners;

import fr.uiytt.ageofempire.ConfigManager;
import fr.uiytt.ageofempire.base.Building;
import fr.uiytt.ageofempire.base.BuildingType;
import fr.uiytt.ageofempire.base.Plot;
import fr.uiytt.ageofempire.game.GameData;
import fr.uiytt.ageofempire.game.GameManager;
import fr.uiytt.ageofempire.game.GameTeam;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.GameMode;
import org.bukkit.Material;
import org.bukkit.Tag;
import org.bukkit.entity.Arrow;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityExplodeEvent;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.projectiles.ProjectileSource;

import java.util.List;
import java.util.UUID;

@SuppressWarnings("unused")
public class GameListener implements Listener {

    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event) {
        GameTeam team = GameManager.getGameInstance().getGameData().getPlayersTeam().get(event.getPlayer().getUniqueId());
        if(GameManager.getGameInstance().getGameData().isGameRunning()) {
            if(team != null) {
                event.getPlayer().setPlayerListName(team.getColor().getTabColor() + event.getPlayer().getDisplayName());
            } else {
                event.getPlayer().setGameMode(GameMode.SPECTATOR);
            }
            return;
        }
        GameTeam.reorganizeTeam();
    }

    @EventHandler
    public void onPlayerLeave(PlayerQuitEvent event) {
        if(GameManager.getGameInstance().getGameData().isGameRunning()) {
            return;
        }
        GameTeam team = GameManager.getGameInstance().getGameData().getPlayersTeam().get(event.getPlayer().getUniqueId());
        if(team != null) {
            team.removePlayer(event.getPlayer().getUniqueId());
        }
    }

    @EventHandler
    public void onBlockPlace(BlockPlaceEvent event) {
        if(!GameManager.getGameInstance().getGameData().isGameRunning()) return;

        ItemMeta itemMeta = event.getItemInHand().getItemMeta();
        if(itemMeta == null) return;

        List<String> lore = itemMeta.getLore();
        if(lore == null || !lore.get(1).contains("AOE")) return;

        Player player = event.getPlayer();

        event.setCancelled(true);

        GameTeam playerTeam = GameManager.getGameInstance().getGameData().getPlayersTeam().get(event.getPlayer().getUniqueId());
        Plot plot = Plot.checkForPlot(playerTeam.getTeamBase(),event.getBlock().getLocation().subtract(0,1,0));
        if(plot == null || !plot.isPlotAvailable()) {
            player.sendMessage("Vous ne pouvez pas construire ici");
            return;
        }

        String buildingName = itemMeta.getLore().get(0).substring(2);
        BuildingType buildingType = BuildingType.getBuildingTypeFromName(buildingName);
        if(buildingType == null) {
            player.sendMessage("Erreur : mauvais item de construction, contactez un admin");
            return;
        }

        Building building = playerTeam.getTeamBase().getBuilds().get(buildingType);
        if(!building.isAvailable()) {
            player.sendMessage("Vous ne pouvez pas construir ça, ce bâtiment existe déjà");
            return;
        }
        if(buildingType.getSize() != plot.getSize()) {
            player.sendMessage("Ce bâtiment doit être construit dans un plot de taille " + buildingType.getSize());
            return;
        }
        event.setCancelled(false);

        event.getBlock().setType(Material.AIR);

        plot.build(playerTeam, buildingType, building);

    }



    @EventHandler
    public void onExplosion(EntityExplodeEvent event) {
        if(!GameManager.getGameInstance().getGameData().isGameRunning()) return;
        event.setYield(0f);
        event.blockList().removeIf(block -> Tag.WOOL.isTagged(block.getType()));
    }

    @EventHandler
    public void onBlockBreak(BlockBreakEvent event) {
        if(!GameManager.getGameInstance().getGameData().isGameRunning()) return;
        Player player = event.getPlayer();
        GameData gameData = GameManager.getGameInstance().getGameData();
        GameTeam gameTeam = gameData.getPlayersTeam().get(player.getUniqueId());

        if(!ConfigManager.getBreakableBlocks().contains(event.getBlock().getType())) {
            event.setCancelled(true);
            return;
        }

        event.setDropItems(false);
        if(event.getBlock().getType() == Material.ANDESITE) {
            gameTeam.getTeamBase().addStone(8);
            gameData.getGold().put(player.getUniqueId(), gameData.getGold().get(player.getUniqueId()) + 10);
            player.sendMessage("+ 8 Stone, +10 Gold");
            player.giveExp(1);
            gameTeam.getTeamBase().updateTeamScoreboard();
        } else if(Tag.LOGS.isTagged(event.getBlock().getType())) {
            gameTeam.getTeamBase().addWood(10);
            gameData.getGold().put(player.getUniqueId(), gameData.getGold().get(player.getUniqueId()) + 10);
            player.sendMessage("+ 10 Bois, +10 Gold");
            player.giveExp(1);
            gameTeam.getTeamBase().updateTeamScoreboard();
        } else if(event.getBlock().getType() == Material.SEA_LANTERN) {
            gameTeam.getTeamBase().addWood(10);
            gameTeam.getTeamBase().addStone(8);
            gameData.getGold().put(player.getUniqueId(), gameData.getGold().get(player.getUniqueId()) + 10);
            player.sendMessage("+8 Stone + 10 Bois, +10 Gold");
            player.giveExp(1);
            gameTeam.getTeamBase().updateTeamScoreboard();
            event.setCancelled(true);
        }

    }

    @EventHandler
    public void onDeath(PlayerDeathEvent event) {
        GameData gamedata = GameManager.getGameInstance().getGameData();
        if(!gamedata.isGameRunning()) return;

        Player deadPlayer = event.getEntity();
        GameTeam playerTeam = GameManager.getGameInstance().getGameData().getPlayersTeam().get(deadPlayer.getUniqueId());

        if(!playerTeam.getTeamBase().isForumAlive()) {
            deadPlayer.setGameMode(GameMode.SPECTATOR);
            return;
        }

        event.getDrops().removeIf(itemStack -> ConfigManager.getDeletedDrops().contains(itemStack.getType()));

        deadPlayer.setGameMode(GameMode.SPECTATOR);

        GameTeam lastTeam = GameManager.getGameInstance().isGameEnd();
        if(lastTeam == null) {
            return;
        }

        Bukkit.getOnlinePlayers().forEach(player -> player.setGameMode(GameMode.SPECTATOR));
        Bukkit.broadcastMessage(ChatColor.GREEN + "" + ChatColor.STRIKETHROUGH + "                                         ");
        Bukkit.broadcastMessage(ChatColor.WHITE + "Victoire de l'equipe " + lastTeam.getColor().getChatColor() + lastTeam.getName() + ChatColor.WHITE + " !!");
        Bukkit.broadcastMessage(ChatColor.GREEN + "" + ChatColor.STRIKETHROUGH + "                                         ");

        GameManager.getGameInstance().stopGame();

    }


    @EventHandler
    public void onChat(AsyncPlayerChatEvent event) {
        GameData gameData = GameManager.getGameInstance().getGameData();
        if(!gameData.isGameRunning()) return;
        GameTeam playerTeam = gameData.getPlayersTeam().get(event.getPlayer().getUniqueId());
        event.setCancelled(true);
        Player player = event.getPlayer();
        if(event.getPlayer().getGameMode() != GameMode.SPECTATOR || playerTeam.getTeamBase().isForumAlive()) {
             if(event.getMessage().charAt(0) == '*') {
                 Bukkit.broadcastMessage(ChatColor.translateAlternateColorCodes('&', "&8&l[" + playerTeam.getColor().getChatColor() + playerTeam.getName() + "&8&l] &e" + player.getDisplayName() + "&7 : " + event.getMessage().substring(1)));
             } else {
                 for(UUID mateUUID : playerTeam.getPlayersUUIDs()) {
                     Player matePlayer = Bukkit.getPlayer(mateUUID);
                     if(matePlayer != null) matePlayer.sendMessage(net.md_5.bungee.api.ChatColor.translateAlternateColorCodes('&',"&8&l[" + playerTeam.getColor().getChatColor() + "EQUIPE&8&l] &e" + player.getDisplayName() + "&7: " + event.getMessage()));
                 }
             }
        } else {
            for(Player onlinePlayer : Bukkit.getOnlinePlayers()) {
                if(onlinePlayer.getGameMode() == GameMode.SPECTATOR) {
                    onlinePlayer.sendMessage(net.md_5.bungee.api.ChatColor.GRAY + "[SPEC] " + player.getDisplayName() + ": " +  event.getMessage());
                }
            }
        }
    }

    @EventHandler
    public void onDamage(EntityDamageByEntityEvent event) {
        if(!(GameManager.getGameInstance().getGameData().isGameRunning()) || GameManager.getGameInstance().getGameData().isPvp()) {
            return;
        }
        if(!(event.getEntity() instanceof Player)) {
            return;
        }
        if(event.getDamager().getType() == EntityType.ARROW) {
            Arrow arrow = (Arrow) event.getDamager();
            ProjectileSource damager = arrow.getShooter();
            if(!(damager instanceof Player)) return;
        } else if(!(event.getDamager() instanceof Player)) return;

        event.setCancelled(true);
    }
}

File: src/main/kotlin/fr/uiytt/ageofempire/listeners/GameListener.kt
package fr.uiytt.ageofempire.listeners

import fr.uiytt.ageofempire.AgeOfEmpire
import fr.uiytt.ageofempire.base.BuildingType.Companion.getBuildingTypeFromName
import fr.uiytt.ageofempire.base.Plot.Companion.checkForPlot
import fr.uiytt.ageofempire.game.GameTeam
import fr.uiytt.ageofempire.game.GameTeam.Companion.reorganizeTeam
import fr.uiytt.ageofempire.game.getGameManager
import fr.uiytt.ageofempire.game.isRunning
import fr.uiytt.ageofempire.getConfigManager
import net.md_5.bungee.api.ChatColor
import org.bukkit.Bukkit
import org.bukkit.GameMode
import org.bukkit.Material
import org.bukkit.Tag
import org.bukkit.block.Block
import org.bukkit.entity.Arrow
import org.bukkit.entity.EntityType
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.entity.EntityExplodeEvent
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.scheduler.BukkitRunnable

class GameListener : Listener {
    @EventHandler
    fun onPlayerJoin(event: PlayerJoinEvent) {
        val team: GameTeam? = getGameManager().gameData.playersTeam[event.player.uniqueId]
        if (getGameManager().isRunning() && team != null) {
            event.player.setPlayerListName(team.color.tabColor.toString() + event.player.displayName)
            return
        } else if (!getGameManager().isRunning()) reorganizeTeam()
    }

    @EventHandler
    fun onPlayerLeave(event: PlayerQuitEvent) {
        if (getGameManager().isRunning()) return
        getGameManager().gameData.playersTeam[event.player.uniqueId]?.removePlayer(event.player.uniqueId)
    }

    @EventHandler
    fun onBlockPlace(event: BlockPlaceEvent) {
        if (!getGameManager().isRunning()) return
        val itemMeta = event.itemInHand.itemMeta ?: return
        val lore = itemMeta.lore
        if (lore == null || !lore[1].contains("AOE")) return
        event.isCancelled = true
        val playerTeam: GameTeam = getGameManager().gameData.playersTeam[event.player.uniqueId]!!
        val plot = checkForPlot(playerTeam.teamBase, event.block.location.subtract(0.0, 1.0, 0.0))

        //Check that plot can be built upon
        if (plot == null || !plot.isPlotAvailable) {
            event.player.sendMessage("Vous ne pouvez pas construire ici")
            return
        }

        //Extract building type from lore of the item
        val buildingName = lore[0].substring(2) //Remove color at the begining
        val buildingType = getBuildingTypeFromName(buildingName)
        if (buildingType == null) {
            event.player.sendMessage("Erreur : mauvais item de construction, contactez un admin")
            return
        }

        //Check that the building is not already constructed
        val building = playerTeam.teamBase.builds[buildingType]
        if (!building!!.isAvailable) {
            event.player.sendMessage("Vous ne pouvez pas construir ça, ce bâtiment existe déjà")
            return
        }
        //Check plot size
        if (buildingType.size != plot.size) {
            event.player.sendMessage("Ce bâtiment doit être construit dans un plot de taille " + buildingType.size)
            return
        }

        event.isCancelled = false
        event.block.type = Material.AIR
        plot.build(playerTeam, buildingType, building)
    }

    @EventHandler
    fun onExplosion(event: EntityExplodeEvent) {
        if (!getGameManager().isRunning()) return
        event.yield = 0f
        event.blockList().removeIf { block: Block -> Tag.WOOL.isTagged(block.type) }
    }

    @EventHandler
    fun onBlockBreak(event: BlockBreakEvent) {
        if (!getGameManager().isRunning()) return
        val gameTeam: GameTeam = getGameManager().gameData.playersTeam[event.player.uniqueId]!!
        if (!getConfigManager().getSetOfBreakableBlocks().contains(event.block.type)) {
            event.isCancelled = true
            return
        }

        event.isDropItems = false
        if (event.block.type == Material.ANDESITE) {
            gameTeam.teamBase.stone += 8
            getGameManager().gameData.addGold(event.player.uniqueId, 10)
            event.player.giveExp(1)
            event.player.sendMessage("+ 8 Stone, +10 Gold")
            gameTeam.teamBase.updateTeamScoreboard()
        } else if (Tag.LOGS.isTagged(event.block.type)) {
            gameTeam.teamBase.wood += 10
            getGameManager().gameData.addGold(event.player.uniqueId, 10)
            event.player.giveExp(1)
            event.player.sendMessage("+ 10 Bois, +10 Gold")
            gameTeam.teamBase.updateTeamScoreboard()
        } else if (event.block.type == Material.SEA_LANTERN) {
            gameTeam.teamBase.wood += 10
            gameTeam.teamBase.stone += 8
            getGameManager().gameData.addGold(event.player.uniqueId, 10)
            event.player.giveExp(1)
            event.player.sendMessage("+8 Stone + 10 Bois, +10 Gold")
            gameTeam.teamBase.updateTeamScoreboard()
            event.isCancelled = true
        }
    }

    @EventHandler
    fun onDeath(event: PlayerDeathEvent) {
        if (!getGameManager().isRunning()) return
        val deadPlayer = event.entity
        val gameTeam: GameTeam = getGameManager().gameData.playersTeam[deadPlayer.uniqueId]!!
        event.drops.removeIf { itemStack: ItemStack -> getConfigManager().getSetOfDeletedDrops().contains(itemStack.type) }
        deadPlayer.gameMode = GameMode.SPECTATOR

        if (gameTeam.teamBase.isForumAlive) {
            object : BukkitRunnable() {
                override fun run() {
                    deadPlayer.gameMode = GameMode.SURVIVAL
                    deadPlayer.teleport(gameTeam.teamBase.spawnTeam)
                }
            }.runTaskLater(AgeOfEmpire.instance, 7*20L)
            return
        }
        Bukkit.broadcastMessage("&f&lLe joueur " + gameTeam.color.chatColor + deadPlayer.displayName + " &f&l est mort définitivement")
        val lastTeam: GameTeam = getGameManager().isGameEnd ?: return
        Bukkit.getOnlinePlayers().forEach { player: Player -> player.gameMode = GameMode.SPECTATOR }
        Bukkit.broadcastMessage(org.bukkit.ChatColor.GREEN.toString() + "" + org.bukkit.ChatColor.STRIKETHROUGH + "                                         ")
        Bukkit.broadcastMessage(org.bukkit.ChatColor.WHITE.toString() + "Victoire de l'equipe " + lastTeam.color.chatColor + lastTeam.name + org.bukkit.ChatColor.WHITE + " !!")
        Bukkit.broadcastMessage(org.bukkit.ChatColor.GREEN.toString() + "" + org.bukkit.ChatColor.STRIKETHROUGH + "                                         ")
        getGameManager().stopGame()
    }

    @EventHandler
    fun onChat(event: AsyncPlayerChatEvent) {
        if (!getGameManager().isRunning()) return
        val gameTeam: GameTeam = getGameManager().gameData.playersTeam[event.player.uniqueId]!!
        event.isCancelled = true
        if (event.player.gameMode != GameMode.SPECTATOR || gameTeam.teamBase.isForumAlive) {
            if (event.message[0] == '*') {
                Bukkit.broadcastMessage(
                    ChatColor.translateAlternateColorCodes(
                        '&',
                        "&8&l[" + gameTeam.color.chatColor + gameTeam.name + "&8&l] &e" + event.player.displayName + "&7 : " + event.message.substring(1)
                    )
                )
            } else {
                gameTeam.playersUUIDs.forEach {
                    Bukkit.getPlayer(it)?.sendMessage(
                        ChatColor.translateAlternateColorCodes(
                            '&',
                            "&8&l[" + gameTeam.color.chatColor + "EQUIPE&8&l] &e" + event.player.displayName + "&7: " + event.message
                        )
                    )
                }
            }
        } else {
            for (onlinePlayer in Bukkit.getOnlinePlayers()) {
                if (onlinePlayer.gameMode == GameMode.SPECTATOR) {
                    onlinePlayer.sendMessage(ChatColor.GRAY.toString() + "[SPEC] " + event.player.displayName + ": " + event.message)
                }
            }
        }
    }

    @EventHandler
    fun onDamage(event: EntityDamageByEntityEvent) {
        if (!getGameManager().isRunning() || getGameManager().gameData.isPvp) return
        if (event.entity !is Player) return
        if (event.damager.type == EntityType.ARROW) {
            val arrow = event.damager as Arrow
            if (arrow.shooter !is Player) return
        } else if (event.damager !is Player) return
        event.isCancelled = true
    }
}